src/server.c

```diff
@@ -5954,25 +5954,26 @@ static sds read_sysfs_line(char *path) {
 
 void linuxTimeWarnings(void) {
 #define PROC_STAT_STIME_IDX 15
-    long long sys_time_ticks, sys_time_ticks_start;
-    unsigned long systime_us, system_hz, test_time_us;
+    unsigned long test_time_us, system_hz;
     struct timespec ts;
     unsigned long long start_us;
+    struct rusage ru_start, ru_end;
 
     /* Check if we're configured to skip this check (useful for slightly faster startup) */
     if (checkIgnoreWarning("SLOW-CLOCKSOURCE"))
         return;
 
     system_hz = sysconf(_SC_CLK_TCK);
 
-    if (!get_proc_stat_ll(PROC_STAT_STIME_IDX, &sys_time_ticks_start))
+    if (getrusage(RUSAGE_SELF, &ru_start) != 0)
         return;
     if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0) {
         return;
     }
     start_us = (ts.tv_sec * 1000000 + ts.tv_nsec / 1000);
 
-    /* clock_gettime() busy loop of 5 times system hz (for a system_hz of 100 this is 50ms).
+    /* clock_gettime() busy loop of 5 times system tick (for a system_hz of 100 this is 50ms)
+     * Using system_hz is required to ensure accurate measurements from getrusage().
```

Comment:
i just tested it on my machine, and it looks like getrusage is much more accurate than `/proc/self/stat`.
even a loop of system_hz / 10 results in accurate readings (not rounded to the next jiffy).
here are some measurements i got
|utime|stime|loops|
|---|---|---|
|669| 334|14065
|336| 671|11475
|1003| 0|12625
|669| 334|14053
|1006| 0|12578
|1003| 0|51488
|670| 335|13639

with the above i could be getting some false positives.
but with a loop for 1 jiffy:

|utime|stime|loops|
|---|---|---|
8570| 1433| 151897
10005| 0| 126577
9132| 494| 115087
9844| 0| 135356
9995| 0| 123566
8888| 1117| 126502

i suppose using a minimum of 1 jiffy or 1000 calls (for slow systems to make sure we have many more clock_gettime readings vs getrusage calls), and setting the threshold to 50% rather than 10% could be good.

Reply:
It still seems to me that the timing we get from `getrusage()` is very coarse and is probably jiffy bound. This is especially problematic for the `xen` clocksource because it's not really 100% kernel or 100% user.
I put some detailed test results here:
https://docs.google.com/spreadsheets/d/141db0CkAQQjMCr1Ivm5HWuZ1C96ly_jHRD6LJYxwMGY/edit#gid=0

Reply:
I discussed it with Yossi, we still fear from the overheads of this check.

the proposed solution is:
1. add some `redis-server --check-system` mode (similar to `--test-memory`), which will do some system checks and report output in stdout and exit code.
2. in normal startup of redis, simply check the configured clock source in `/proc` and report a warning if we see `Xen`. (the warning should mention the `--check-system` feature, and can be disabled)

Reply:
Done. Please re-review.